python_modules/dagster/dagster/_core/storage/asset_check_execution_record.py

```diff
@@ -101,3 +121,91 @@ def from_db_row(cls, row) -> "AssetCheckExecutionRecord":
+                return (
+                    AssetCheckExecutionResolvedStatus.EXECUTION_FAILED
+                    if run.status == DagsterRunStatus.FAILURE
+                    else AssetCheckExecutionResolvedStatus.SKIPPED
+                )
+            else:
+                return AssetCheckExecutionResolvedStatus.IN_PROGRESS
+        else:
+            check.failed(f"Unexpected status {self.status}")
+
+    def targets_latest_materialization(self, loading_context: LoadingContext) -> bool:
+        from dagster._core.storage.event_log.base import AssetRecord
+
+        resolved_status = self.resolve_status(loading_context)
+        if resolved_status == AssetCheckExecutionResolvedStatus.IN_PROGRESS:
+            # all in-progress checks execute against the latest version
+            return True
+
+        asset_record = AssetRecord.blocking_get(loading_context, self.key.asset_key)
+        latest_materialization = (
+            asset_record.asset_entry.last_materialization_record if asset_record else None
+        )
+        if not latest_materialization:
+            # no previous materialization, so it's executing against the lastest version
+            return True
+
+        latest_materialization_run_id = latest_materialization.event_log_entry.run_id
+        if latest_materialization_run_id == self.run_id:
+            # part of the same run
+            return True
+
+        if resolved_status in [
+            AssetCheckExecutionResolvedStatus.SUCCEEDED,
+            AssetCheckExecutionResolvedStatus.FAILED,
+        ]:
+            evaluation = check.not_none(self.evaluation)
+            if not evaluation.target_materialization_data:
+                # check ran before the materialization was created
+                return False
+            else:
+                # check matches the latest materialization id
+                return (
+                    evaluation.target_materialization_data.storage_id
+                    == latest_materialization.storage_id
+                )
+        elif resolved_status in [
+            AssetCheckExecutionResolvedStatus.EXECUTION_FAILED,
+            AssetCheckExecutionResolvedStatus.SKIPPED,
+        ]:
+            # the evaluation didn't complete, so we don't have target_materialization_data, so check if
+            # the check's run executed after the materializations as a fallback
+            latest_materialization_run_record = RunRecord.blocking_get(
+                loading_context, latest_materialization_run_id
+            )
+            check_run_record = RunRecord.blocking_get(loading_context, self.run_id)
+            return bool(
+                latest_materialization_run_record
+                and check_run_record
+                and check_run_record.create_timestamp
+                > latest_materialization_run_record.create_timestamp
+            )
+        else:
+            check.failed(f"Unexpected check status {resolved_status}")
```

Comment:
there will be non-gql callsites further upstack (in DA land)

Reply:
If we wrote async versions now since they could be called that way in their current call-site, could we `asyncio.run` them from DA? 

If we believe async is the end game im just thinking through how we can progress in that direction

Reply:
my concern was basically that this would mean that these async calls would not share a cache with the sync versions, and so we'd miss out on some perf benefits in DA -- agree in principle though.

part of me just wants to churn through the more targeted refactors here and then we can take a more birds-eye view of what macro changes we can make to get stuff to execute more asynchronously

Reply:
Ok, you have more context on the full suite of changes so ill defer to your judgement